### ISSUE TITLE
Questions regarding Forecasting Model (TFT) training and evaluation

### ISSUE THREAD

**Issue by cosun1210** [spaceml-org/karman#9]
Dear Professor Giacomo Acciarini,

Greetings! This is Sun Yuxuan. We have previously been in contact via email regarding the Karman project. Thank you again for your guidance and the help you've provided so far.

I am still working on reproducing the training for your forecasting models, and I've encountered a few new questions where I would greatly appreciate your insights:

1.  **Training Target Type:** For the provided pretrained forecasting models like `ts_karman_model_tft_ss32...` (MAPE 15.906%) etc., was the target type used during training `log_density` or `log_exp_residual`? (This is crucial for me to correctly handle the model's output.)

2.  **Number of Static Features:** I noticed that in `density_models.py`, the `ForecastingModel` class initializes with `num_static_features` defaulting to 8. However, the default static normalization dictionary (`_normalization_dict`) in the library seems to contain around 14 physically relevant static features (excluding coordinates, time, and the target variable itself). Additionally, the Nowcasting model (`SimpleNetwork` in `train_nowcasting_karman.py`) uses an input dimension of 18. Does the TFT Forecasting model indeed expect only 8 specific static features as input? If so, could you please clarify which 8 features they are and why the number differs from the Nowcasting model or the features present in `_normalization_dict`?

3.  **Library Function Modification:** When attempting to implement my own training/evaluation pipeline for the Forecasting model, is it expected that I might need to modify core library functions (e.g., functions within `util.py`, `density_models.py`, etc.)? Does the current code in the GitHub repository support custom Forecasting model training/evaluation workflows 'out-of-the-box', or is there a recommended, more robust way to customize this (perhaps through subclassing or configuration)?

4.  **Data Time Range:** For the provided pretrained forecasting models like `ts_karman_model_tft_ss32...` (MAPE 15.906%), was the end date of the data used for training also `2024-05-31 23:59:32`?

Thank you once again for your time and the valuable Karman project.

Best regards,
Sun Yuxuan

**Comment by cosun1210** [comment]
After researching your model, I noticed that it appears to expect **8 static features** and **1 future feature**. However, the current library functions seem to provide **18 static features** and **10 future features**.

Below is the output from my training inspection, where I checked the static and future features being used:

```python
--- Features Used by KarmanDataset ---
Static Features ('instantaneous_features') (18):
  tudelft_thermo__altitude__[m], tudelft_thermo__latitude__[deg], celestrack__ap_average__, space_environment_technologies__f107_obs__
  space_environment_technologies__f107_average__, space_environment_technologies__s107_obs__, space_environment_technologies__s107_average__, space_environment_technologies__m107_obs__
  space_environment_technologies__m107_average__, space_environment_technologies__y107_obs__, space_environment_technologies__y107_average__, JB08__d_st_dt__[K]
  tudelft_thermo__longitude__[deg]_sin, tudelft_thermo__longitude__[deg]_cos, all__day_of_year__[d]_sin, all__day_of_year__[d]_cos
  all__seconds_in_day__[s]_sin, all__seconds_in_day__[s]_cos

Future Features ('msise' source) (10):
  nrlmsise00__lat_90.0_lon_0.0__[kg/m**3], nrlmsise00__lat_51.1_lon_137.5__[kg/m**3], nrlmsise00__lat_33.7_lon_-85.0__[kg/m**3]
  nrlmsise00__lat_19.5_lon_52.5__[kg/m**3], nrlmsise00__lat_6.4_lon_-170.0__[kg/m**3], nrlmsise00__lat_-6.4_lon_-32.5__[kg/m**3]
  nrlmsise00__lat_-19.5_lon_105.0__[kg/m**3], nrlmsise00__lat_-33.7_lon_-117.4__[kg/m**3], nrlmsise00__lat_-51.1_lon_20.1__[kg/m**3]
  nrlmsise00__lat_-90.0_lon_180.0__[kg/m**3]
--------------------------------------
```

I have the following questions:

1. Can you confirm whether the `ts_karman_model_tft_ss...torch` model was indeed trained with a configuration of **8 static features and 1 future feature**?

2. If it uses **8 static features**, which specific ones are they? My guess is `altitude`, `latitude`, `lon_sin`, `lon_cos`, `doy_sin`, `doy_cos`, `sid_sin`, and `sid_cos`, excluding space weather indices. Is this guess correct?

3. What exactly is the **1 future feature**? Is it the density prediction value from **NRLMSISE00** at each sample point’s corresponding location and time? If so, how is this value generated and provided to the training process? Is it sourced from the `nrlmsise00_time_series.csv` file (which includes data for 10 grid points)? Is this file related to the training of this model?

4. If I want to use the Karman library code to train a model with the **8/1 configuration**, is there a recommended configuration method (e.g., passing a specific `features_to_exclude_thermo` list when initializing `KarmanDataset`)? Are there any data processing suggestions, especially on how to obtain just that **1 future feature** from the MSISE data?

Thank you for open-sourcing this project, and I apologize for any inconvenience caused. I hope you can clarify my doubts. Thank you.

**Comment by Sceki** [comment]
Hi @cosun1210 ,

Thank you for your interest & questions!

Concerning your first message:

1. to understand how to use those models for inference you can have a look here: https://github.com/spaceml-org/karman/blob/main/karman/density_models.py#L190C1-L193C36.. it should be `log_density` (judging from these lines)
2. to understand how to deal with the static/non-static features and their length, you can have a look here: https://github.com/spaceml-org/karman/blob/main/karman/util.py#L166 for the forecasting, and here: https://github.com/spaceml-org/karman/blob/main/karman/util.py#L79, for the nowcasting...this should clarify your doubts/questions
3. no it is not expected. As you see in this tutorial: `https://github.com/spaceml-org/karman/blob/main/docs/notebooks/tutorial_forecast.ipynb`, you should be able to perform forecasting without substantially modifying anything in the library
4. good question.. I think it should be: `max_date: 2024-05-31 23:59:32, min_date: 2000-07-29 00:59:47`

**Comment by Sceki** [comment]
As for the second question:

> After researching your model, I noticed that it appears to expect 8 static features and 1 future feature. However, the current library functions seem to provide 18 static features and 10 future features.

I think I replied above.. you can directly see in those scripts how to do it and the number of static/non-static features (directly taken from the normalization dictionaries).

And for the follow-up questions:

1. it was trained with the number of features consistent with the scripts I mentioned above. You can directly check it from there (I now don't remember by heart and would need to check myself)
2. same as above: the procedure to get the time series data, normalize it, choose them etc. is all open and you can check it following the nowcasting/forecasting tutorials and inspecting the `density_models.py` file and the functions in the `util.py` if interested
3. same reply as above
4. I don't think you should assume 8/1 configuration. The inputs used to train is the same that I specify in the forecasting scripts/tutorials, you should not modify any parts of the library to work with that

**Comment by cosun1210** [comment]
Dear Professor Giacomo Acciarini,

Thank you again for your previous detailed guidance. I've made some new progress in attempting to reproduce your forecasting model (specifically, ts_karman_model_tft_ss64_heads4_lag10000_resolution100_valid_mape_15.059_params_1074865.torch – as I mentioned before, I believe its name in the GitHub project, ss65_heads4, might be a typo for ss64). However, this has also led to some new questions where I would greatly appreciate your further clarification.

1.  **Successful Parameter Count Match & TFT Configuration Discovery**:
    * I wrote a small script to test TFT parameter counts. I found that when the TFT model's core hyperparameters are set to `state_size=64`, `attention_heads=4`, `lstm_layers=2`, `dropout=0.01`, and its `data_props` are configured as `{ 'num_static_numeric': 8, 'num_historical_numeric': 25, 'num_future_numeric': 1 }`, the calculated total trainable parameters is exactly **1,074,865**, which perfectly matches the parameter count in your model's filename.
    * I also noted that this `data_props` configuration (i.e., 8 static, 25 historical, and 1 future feature) is identical to the `data_props` used by the `_tft_configuration` within the `karman.density_models.ForecastingModel` class when it's initialized with its default parameters (i.e., `num_static_features=8`, `num_future_numeric=1`).

2.  **Reconciling with Your Previous Advice and the Tutorial**:
    * You previously advised me that I "shouldn't assume an 8/1 configuration" and pointed out that "the inputs used to train is the same that I specify in the forecasting scripts/tutorials, you should not modify any parts of the library to work with that." However, your model does appear to use 8 static features and 1 future feature, while the library (which I haven’t modified) indeed outputs 18 static features and 10 future features — this is the core issue I’m currently facing.
    * I've carefully reviewed `tutorial_forecast.ipynb`. When instantiating `ForecastingModel` in the tutorial, the `num_static_features` and `num_future_numeric` parameters are indeed not overridden, so they take their default values of 8 and 1, respectively. The `features_to_exclude_thermo` list defined in the tutorial also explicitly excludes space weather indices, which aligns with having 8 static features (i.e., only altitude, latitude, and the sin/cos forms of longitude, day-of-year, and second-in-day).

3.  **My Current Main Points of Confusion and Specific Questions**:

    * **a) Regarding Static Features (8 vs. my previous understanding of 18)**:
        * Given the parameter count match, the default behavior of the `ForecastingModel` class, and the feature exclusion method in the tutorial, does this imply that the `ts_..._params_1074865.torch` model was indeed trained with these 8 core static features as input (meaning space weather indices, etc., were **not** included as static inputs)? This differs from my previous training script, which used the `KarmanDataset`'s default output of 18 static features (including space weather indices).

    * **b) Regarding Future Feature Handling (TFT config `num_future_numeric: 1` vs. 10 MSISE grid point data source)**:
        * I noticed that in your project's `scripts/input_data_prep/run_nrlmsise00_time_series.py` script, the `nrlmsise00_time_series.csv` file is generated by calculating NRLMSISE00 density time series for 10 fixed latitude/longitude points (at an altitude of 350km). Therefore, when the tutorial and the `karman.util.get_ts_data` function prepare MSISE future feature data from this CSV, they indeed obtain a tensor whose last dimension is 10 (corresponding to these 10 grid points).
        * If the `ts_..._params_1074865.torch` model was trained with its TFT `data_props` configured for `num_future_numeric: 1`, how does the TFT model handle this 10-dimensional `future_ts_numeric` input? For instance, does the `tft_torch` library or the model, when processing, default to selecting only one specific feature (e.g., the first one) from these 10 MSISE grid point data as the effective future feature input? Or is there another aggregation/selection mechanism to process the 10-dimensional information into the single dimension expected by the TFT configuration?

    * **c) Regarding Historical Features (25)**:
        * The `ForecastingModel` class's `compute_data_props` method, by default, calculates 25 historical features (including all OMNI sources [6+4+3], SOHO [2], and all 10 MSISE grid point data as historical inputs). Is my understanding that the `ts_..._params_1074865.torch` model also uses these 25 historical features correct?

4.  **Current Evaluation Results**:
    * With the above adjustments (TFT configured for Static=8, Historical=25, Future=1; selecting 8 core static features; and selecting the first of the 10 MSISE future features), my new evaluation script can now successfully load your `ts_..._params_1074865.torch` model.
    * However, when using the original `.pk` normalization files provided within your library, the MAPE I obtain on the validation set is approximately 19.9%, which still differs from the 15.059% noted in the model's filename. I suspect this remaining discrepancy might be related to subtle differences in the data files, the precise definition of the validation set split, or perhaps a more nuanced method for deriving the single future MSISE feature (if not simply by taking the first one).

Could you please provide some clarification on points 3a, 3b, and 3c above? Specifically, confirming whether the static features are indeed the 8 core ones, and how the single future MSISE feature (if the TFT is configured for 1) is obtained from the 10-dimensional MSISE data for the training of the `ts_..._params_1074865.torch` model, would be immensely helpful for me to accurately reproduce your workflow and results.

Thank you once again for your time and invaluable assistance!

Best regards,
Sun Yuxuan

**Comment by Sceki** [comment]
Hi there,

- 3a) As I said above, you should not assume any configuration. What I meant is that the configuration can be seen by using those scripts. When calling: `forecasting_model.compute_data_props()` from the tutorial for instance I can see that we have: 
```Python
{'num_historical_numeric': 25,
 'num_static_numeric': 8,
 'num_future_numeric': 1}
```
so, all good if you also obtain that. I encourage you not to change any base util/class in the library (as you suggested in previous messages), else you might get reproducibility issues
- 3b) you can have a look here to derive the dimensionality of the future ts and how it's constructed: https://github.com/spaceml-org/karman/blob/main/karman/util.py#L227
- 3c) yes, 25 historical features
- 4)-> indeed you seem to still have discrepancy: this could be in how you prepared the data.. in previous messages you mentioned modifying the core functions of the library, this might also cause the divergence.. 

Good luck..
Giacomo

**Comment by Sceki** [comment]
Closing this for now.. will reopen if needed :)
